File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0);cout.tie(0);ios::sync_with_stdio(false);
    int k, n; cin >> k >> n;
    vector<ll> line(k);
    ll low = 1, high = 0;
    for(int i = 0; i < k; i++){
        cin >> line[i];
        if(high < line[i]) high = line[i];
    }
    while(low <= high){
        ll mid = (low+high)/2;
        ll total = 0;
        for(auto l : line){
            total += l/mid;
        }
        if(total >= n) low = mid+1;
        else high = mid-1;
    }
    cout << high;
    return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

def main():
    k, n = map(int, input().split())
    line = tuple(int(input()) for _ in range(k))
    low = 1; high = max(line)
    while low <= high:
        mid = (low+high)//2
        total = sum(l//mid for l in line)
        if total >= n: low = mid+1
        else: high = mid-1
    print(high)
    return 0
if __name__ == '__main__':
    sys.exit(main())